### PR TITLE
Disable linux/ppc64le compilation by default

### PR DESCRIPTION
**Pull request by ixdy**
Work-around for #30384.

I'm still testing this locally to see if it actually works. The build is slow. (PR Jenkins won't tell us whether this fixes ppc.)

cc @Random-Liu @spxtr @david-mcmahon @luxas

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30659)

### CODE REVIEW

**Comment by Random-Liu** [comment]
@ixdy Thanks for the fix! :)

**Comment by spxtr** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30659#-:-KPFqFTqo-S7MH1WlWFJ:bnfp4nl)*

**Comment by ixdy** [comment]
This doesn't fix the build. kube-apiserver is still broken:
```
hyperkube and kubemark builds for ppc64le are disabled. Creating dummy text file instead.
hyperkube and kubemark builds for ppc64le are disabled. Creating dummy text file instead.
# k8s.io/kubernetes/cmd/kube-apiserver
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.PersistentVolumeReclaimPolicy.CodecEncodeSelf: relocation for runtime.convT2E+220 is t
oo big: -33558660
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeStatus).CodecDecodeSelf: relocation for runtime.convI2E+592 is too b
ig: -33559048
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeList).CodecDecodeSelf: relocation for runtime.convI2E+592 is too big
: -33565120
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaim).CodecDecodeSelf: relocation for runtime.convI2E+592 is too bi
g: -33571896
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaim).codecDecodeSelfFromMap: relocation for runtime.writebarrierpt
r+860 is too big: -33558476
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaim).codecDecodeSelfFromMap: relocation for runtime.writebarrierpt
r+1716 is too big: -33559332
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaim).codecDecodeSelfFromArray: relocation for runtime.writebarrier
ptr+1692 is too big: -33561044
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaim).codecDecodeSelfFromArray: relocation for runtime.writebarrier
ptr+1848 is too big: -33561200
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimList).CodecDecodeSelf: relocation for runtime.convI2E+592 is to
o big: -33578944
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimList).codecDecodeSelfFromMap: relocation for runtime.writebarri
erptr+860 is too big: -33565524
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimList).codecDecodeSelfFromMap: relocation for runtime.writebarri
erptr+1716 is too big: -33566380
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimList).codecDecodeSelfFromArray: relocation for runtime.writebar
rierptr+1592 is too big: -33567992
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimList).codecDecodeSelfFromArray: relocation for runtime.writebar
rierptr+1748 is too big: -33568148
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimSpec).CodecDecodeSelf: relocation for runtime.convI2E+592 is to
o big: -33585848
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimSpec).codecDecodeSelfFromMap: relocation for runtime.newobject+
796 is too big: -33575244
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimSpec).codecDecodeSelfFromMap: relocation for runtime.writebarri
erptr+1072 is too big: -33572640
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimSpec).codecDecodeSelfFromMap: relocation for runtime.writebarrierptr+1524 is too big: -33573092
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimSpec).codecDecodeSelfFromArray: relocation for runtime.writebarrierptr+1180 is too big: -33574484
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimSpec).codecDecodeSelfFromArray: relocation for runtime.newobject+1356 is too big: -33577540
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimSpec).codecDecodeSelfFromArray: relocation for runtime.writebarrierptr+1632 is too big: -33574936
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api.(*PersistentVolumeClaimStatus).CodecDecodeSelf: relocation for runtime.convI2E+592 is too big: -33591976
/usr/local/go/pkg/tool/linux_amd64/link: too many errors
# k8s.io/kubernetes/cmd/kube-controller-manager
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.mapiterinit+216 is too big: -33568568
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.mapiternext+412 is too big: -33567748
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.mapaccess1_faststr+620 is too big: -33558844
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+2212 is too big: -33554468
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+2428 is too big: -33554684
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+2644 is too big: -33554900
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+2860 is too big: -33555116
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+3076 is too big: -33555332
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+3292 is too big: -33555548
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+3508 is too big: -33555764
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+3956 is too big: -33556212
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+4312 is too big: -33556568
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodSpec).String: relocation for runtime.convT2E+4528 is too big: -33556784
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodStatus).String: relocation for runtime.convT2E+176 is too big: -33559992
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodStatus).String: relocation for runtime.convT2E+392 is too big: -33560208
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodStatus).String: relocation for runtime.convT2E+748 is too big: -33560564
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodStatus).String: relocation for runtime.convT2E+964 is too big: -33560780
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodStatus).String: relocation for runtime.convT2E+1180 is too big: -33560996
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodStatus).String: relocation for runtime.convT2E+1396 is too big: -33561212
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodStatus).String: relocation for runtime.convT2E+1844 is too big: -33561660
k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/pkg/api/v1.(*PodTemplateList).String: relocation for runtime.convT2E+392 is too big: -33565616
/usr/local/go/pkg/tool/linux_amd64/link: too many errors
```

**Comment by luxas** [comment]
@ixdy I think we have to uncomment ppc64le completely for now :cry: (at least for this alpha release)

We should instead talk about if we can accept building ppc64le separate with the patched golang linker.
cc @laboger @pensu @bpradipt

**Comment by Pensu** [comment]
I was able to build with patched golang, would it be possible to have a separate build for ppc64le with patched golang?

**Comment by spxtr** [comment]
I guess now is a good time to add a separate CI job for other architectures.

**Comment by ixdy** [comment]
I don't think we have an easy way to make our CI or release infrastructure use a non-standard golang (especially not for just one arch), so I'm not sure how feasible it is...

**Comment by spxtr** [comment]
It *should* be as simple as building the patched golang into a special image that we use for those architectures only, but unfortunately it won't be that simple.

**Comment by luxas** [comment]
@david-mcmahon  ^

**Comment by ixdy** [comment]
reworked to disable `linux/ppc64le` compilation by default. :(

**Comment by ixdy** [comment]
The env var override doesn't work correctly in dockerized builds. Working on a fix...

**Comment by ixdy** [comment]
OK, I think this works finally.

I removed the special-casing for hyperkube since I'm guessing fixing the rest of ppc will fix it too. Seem reasonable @luxas?

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -668,7 +666,7 @@ kube::golang::build_binaries() {
     kube::golang::build_kube_toolchain
 
     if [[ "${parallel}" == "true" ]]; then
-      kube::log::status "Building go targets for ${platforms[@]} in parallel (output will appear in a burst when complete):" "${targets[@]}"
+      kube::log::status "Building go targets for {${platforms[*]}} in parallel (output will appear in a burst when complete):" "${targets[@]}"
```
Comment (ixdy) [inline hack/lib/golang.sh]: this change wasn't really necessary, but now the platforms are all listed on one line, which is nice. e.g.

```
+++ [0816 17:05:44] Building go targets for {linux/amd64 linux/arm linux/arm64} in parallel (output will appear in a burst when complete):
    cmd/kube-dns
    cmd/kube-proxy
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kubelet
    cmd/kubemark
    cmd/hyperkube
    federation/cmd/federation-apiserver
    federation/cmd/federation-controller-manager
    plugin/cmd/kube-scheduler
```
and
```
+++ [0816 17:02:26] Building go targets for {linux/amd64 linux/arm linux/arm64 linux/ppc64le} in parallel (output will appear in a burst when complete):
    cmd/kube-dns
    cmd/kube-proxy
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kubelet
    cmd/kubemark
    cmd/hyperkube
    federation/cmd/federation-apiserver
    federation/cmd/federation-controller-manager
    plugin/cmd/kube-scheduler
```

**Comment by luxas** [comment]
@k8s-bot unit test this please github issue: #IGNORE

**Comment by luxas** [comment]
LGTM

cc @spxtr

**Comment by david-mcmahon** [comment]
ping.  Are we waiting on this for anything else?

**Comment by ixdy** [comment]
was waiting to see if anyone else had opinions.

hearing none, I'm gonna tag it.

**Comment by ixdy** [comment]
also marking P2 so we can hopefully reenable `make release` builds before we break things again.

**Comment by spxtr** [comment]
Pretty strange. This is merged, but gubernator/mergebot/the github UI didn't get the memo.

**Comment by ixdy** [comment]
yeah, that's weird.

**Pull request merged**